File: apps/api-rs/src/api/tenant_login.rs
use axum::extract::{Query, State};
use axum::routing::get;
use axum::{Json, Router};
use serde::{Deserialize, Serialize};
use serde_json::Value;
use sqlx::Row;

use crate::error::AppError;
use crate::state::AppState;

const MAX_TENANT_SLUG_LENGTH: usize = 63;

#[derive(Debug, Deserialize)]
pub struct TenantLoginQuery {
    tenant: String,
}

#[derive(Debug, Serialize, PartialEq)]
pub struct TenantLoginBranding {
    #[serde(skip_serializing_if = "Option::is_none")]
    kicker: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    headline: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    description: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    welcome: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    tone: Option<String>,
}

#[derive(Debug, Serialize, PartialEq)]
pub struct TenantLoginContext {
    name: String,
    slug: String,
    branding: TenantLoginBranding,
}

pub fn router() -> Router<AppState> {
    Router::new()
        .route("/api/auth/tenant-context", get(get_tenant_login_context))
        .route("/api/v1/auth/tenant-context", get(get_tenant_login_context))
}

async fn get_tenant_login_context(
    State(state): State<AppState>,
    Query(query): Query<TenantLoginQuery>,
) -> Result<Json<TenantLoginContext>, AppError> {
    let slug = normalize_tenant_slug(&query.tenant)?;
    let pool = state.db_pool.as_ref().ok_or_else(|| {
        AppError::Internal(anyhow::anyhow!(
            "tenant login context requires a database connection"
        ))
    })?;
    let row = sqlx::query(
        r#"
        SELECT name, slug, branding
        FROM tenant
        WHERE lower(slug) = lower($1)
          AND status = 'active'
        "#,
    )
    .bind(&slug)
    .fetch_optional(pool)
    .await?
    .ok_or_else(tenant_login_not_found)?;
    let branding = row.get::<Value, _>("branding");

    Ok(Json(TenantLoginContext {
        name: row.get("name"),
        slug: row.get("slug"),
        branding: parse_login_branding(&branding),
    }))
}

fn normalize_tenant_slug(value: &str) -> Result<String, AppError> {
    let slug = value.trim().to_ascii_lowercase();
    let valid = (2..=MAX_TENANT_SLUG_LENGTH).contains(&slug.len())
        && slug
            .bytes()
            .all(|byte| byte.is_ascii_alphanumeric() || byte == b'-');
    if !valid {
        return Err(tenant_login_not_found());
    }
    Ok(slug)
}

fn tenant_login_not_found() -> AppError {
    AppError::NotFound {
        code: "TENANT_LOGIN_NOT_FOUND".to_string(),
        message: "企业登录入口不存在或暂不可用".to_string(),
    }
}

fn parse_login_branding(value: &Value) -> TenantLoginBranding {
    TenantLoginBranding {
        kicker: branding_text(value, "login_kicker", 32),
        headline: branding_text(value, "login_headline", 64),
        description: branding_text(value, "login_description", 120),
        welcome: branding_text(value, "login_welcome", 40),
        tone: branding_tone(value),
    }
}

fn branding_text(value: &Value, key: &str, max_chars: usize) -> Option<String> {
    let text = value.get(key)?.as_str()?.trim();
    if text.is_empty() {
        return None;
    }
    Some(text.chars().take(max_chars).collect())
}

fn branding_tone(value: &Value) -> Option<String> {
    let tone = value.get("login_tone")?.as_str()?.trim();
    match tone {
        "violet" | "azure" | "jade" | "amber" | "rose" => Some(tone.to_string()),
        _ => None,
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use serde_json::json;

    #[test]
    fn normalizes_public_tenant_slug() {
        assert_eq!(normalize_tenant_slug("  Acme-01 ").unwrap(), "acme-01");
        assert!(normalize_tenant_slug("a").is_err());
        assert!(normalize_tenant_slug("acme_01").is_err());
        assert!(normalize_tenant_slug("租户").is_err());
    }

    #[test]
    fn exposes_only_supported_login_branding() {
        let branding = parse_login_branding(&json!({
            "login_kicker": "  可信知识，持续生长  ",
            "login_headline": "让经验成为共同的判断依据",
            "login_description": "连接团队文档与业务上下文。",
            "login_welcome": "欢迎回到知识中枢",
            "login_tone": "jade",
            "secret": "must-not-leak"
        }));
        assert_eq!(branding.kicker.as_deref(), Some("可信知识，持续生长"));
        assert_eq!(branding.tone.as_deref(), Some("jade"));
    }

    #[test]
    fn rejects_unknown_tone_and_empty_copy() {
        let branding = parse_login_branding(&json!({
            "login_kicker": " ",
            "login_tone": "neon"
        }));
        assert_eq!(branding.kicker, None);
        assert_eq!(branding.tone, None);
    }
}

File: apps/api-bun/src/api/tenant_login.ts
// 移植自 apps/api-rs/src/api/tenant_login.rs
import { Hono } from 'hono';
import { AppError } from '../errors.ts';
import type { AppEnv } from '../http/types.ts';

const MAX_TENANT_SLUG_LENGTH = 63;

interface TenantLoginBranding {
  kicker?: string | null; headline?: string | null; description?: string | null;
  welcome?: string | null; tone?: string | null;
}
interface TenantLoginContext { name: string; slug: string; branding: TenantLoginBranding; }

export function tenantLoginRouter(): Hono<AppEnv> {
  const router = new Hono<AppEnv>();
  router.get('/api/auth/tenant-context', getTenantLoginContextHandler);
  router.get('/api/v1/auth/tenant-context', getTenantLoginContextHandler);
  return router;
}

async function getTenantLoginContextHandler(c: import('hono').Context<AppEnv>) {
  const state = c.get('appState');
  const tenant = c.req.query('tenant') ?? '';
  const slug = normalizeTenantSlug(tenant);
  const sql = state.sql;
  if (!sql) throw AppError.internal('tenant login context requires a database connection');
  const rows = await sql`
    SELECT name, slug, branding
    FROM tenant
    WHERE lower(slug) = lower(${slug})
      AND status = 'active'
  `;
  const row = rows[0];
  if (!row) throw tenantLoginNotFound();
  return c.json({
    name: String(row.name), slug: String(row.slug),
    branding: parseLoginBranding((row.branding ?? {}) as Record<string, unknown>),
  } satisfies TenantLoginContext);
}

export function normalizeTenantSlug(value: string): string {
  const slug = value.trim().toLowerCase();
  const valid = slug.length >= 2 && slug.length <= MAX_TENANT_SLUG_LENGTH
    && [...slug].every((ch) => /[a-z0-9]/.test(ch) || ch === '-');
  if (!valid) throw tenantLoginNotFound();
  return slug;
}

function tenantLoginNotFound(): AppError {
  return AppError.notFound('TENANT_LOGIN_NOT_FOUND', '企业登录入口不存在或暂不可用');
}

export function parseLoginBranding(value: Record<string, unknown>): TenantLoginBranding {
  const branding: TenantLoginBranding = {};
  const kicker = brandingText(value, 'login_kicker', 32);
  if (kicker !== null) branding.kicker = kicker;
  const headline = brandingText(value, 'login_headline', 64);
  if (headline !== null) branding.headline = headline;
  const description = brandingText(value, 'login_description', 120);
  if (description !== null) branding.description = description;
  const welcome = brandingText(value, 'login_welcome', 40);
  if (welcome !== null) branding.welcome = welcome;
  const tone = brandingTone(value);
  if (tone !== null) branding.tone = tone;
  return branding;
}

function brandingText(value: Record<string, unknown>, key: string, maxChars: number): string | null {
  const raw = value[key];
  if (typeof raw !== 'string') return null;
  const text = raw.trim();
  if (text.length === 0) return null;
  return [...text].slice(0, maxChars).join('');
}

function brandingTone(value: Record<string, unknown>): string | null {
  const raw = value.login_tone;
  if (typeof raw !== 'string') return null;
  const tone = raw.trim();
  return ['violet', 'azure', 'jade', 'amber', 'rose'].includes(tone) ? tone : null;
}
